### PR TITLE
Implement functionality to navigate from test method to validation file and back

### PR DESCRIPTION
**Pull request by kevin-keul**
See issue https://github.com/cronn/validation-files-comparison-intellij-plugin/issues/18

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,6 +43,10 @@ When renaming a test with validation files the validation and output files will
 Simply use the rename refactoring of intellij (`Shift+F6` / `⇧F6`) and apply the suggested renaming as wished.
 To disable this feature simply trigger the keybinding for the renaming twice and deselect "Rename validation and output files".
 
+### Jump to Validation File or Code
+
+Similar to IntelliJ's build-in action for switching between tests and implementations, it is possible to directly jump into the according validation file from a test and back. Simply press the default shortcut `Alt+Shift+F` /`⌥⇧F` and the according file is opened at the relevant position. If there are multiple possible targets, a selection window with all targets is presented. This action can be used from within a test method as well as for the whole test class. When used from a validation file, it finds the deepest possible code target to jump to, also regarding nested test classes and methods with multiple validation files.
+
```
Comment (aykborstelmann) [inline README.md:49]: Adjust changelog, as documented in `README.md`

File: src/test/java/de/cronn/validation_files_diff/impl/GoToValidationFileTest.java [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+			// folders
+			"<!C!>, ~o/Class", // output folder
+			"<!C!>, ~v/Class", // validation folder
+
+			// class
+			"<!C!>, ~v/Class/", // as directory
+			"<!C!>, ~v/Class2", // as prefix
+			"<!C!>, ~v/Class2/", // prefixed class as directory
+			"<!C!>, ~v/Class.txt", // with extension
+			"<!C!>, ~v/Class_other.txt", // with continuation
```
Comment (pfhartma) [inline src/test/java/de/cronn/validation_files_diff/impl/GoToValidationFileTest.java]: As discussed in person, our project has use cases that may not be covered right now:

Example are validation files
* `data/test/validation/SepaPrenotificationIntegrationTest/Membership/testDownloadSepaPrenotification_noSepaPaymentMethod.json`
* `data/test/validation/SepaPrenotificationIntegrationTest/Membership/testDownloadSepaPrenotification_noSepaPaymentMethod_logging.txt`

for a test method `SepaPrenotificationIntegrationTest.Membership#testDownloadSepaPrenotification_noSepaPaymentMethod`

The important part is the custom suffix in the valildation file name that extends the method name. It is used to support multiple validation files per test in a lightweight way. Such a suffix may also come from parameterized tests.

The jump to the validation file works fine, but when jumping back to the method, it reaches the enclosing class.

**Comment by pfhartma** [comment]
Update: test works on the current version (9c165260) of the branch

~~On my system, the tests do not run successfully. (It works for the master branch though)~~
```
$ java -version
openjdk version "17.0.4" 2022-07-19
```
```
./gradlew test
```

failing tests:
```
UnknownClass. executionError
ValidationFileAutomaticRenamerTest. testRenameClass_renamesAllValidationFiles
ValidationFileAutomaticRenamerTest. testRenameClass_withDirectoryJoiningStrategy_renamesOnlyApplicableValidationFiles
ValidationFileAutomaticRenamerTest. testRenameMethod_renamesAllValidationFiles
ValidationFileAutomaticRenamerTest. testRenameMethod_withDirectoryJoiningStrategy_renamesOnlyApplicableValidationFiles
```

```
com.intellij.testFramework.TestLoggerFactory$TestLoggerAssertionError: Memory leak detected: 'newDisposable' of class com.intellij.openapi.util.Disposer$1 is registered in Disposer but wasn't disposed.
Register it with a proper parentDisposable or ensure that it's always disposed by direct Disposer.dispose call.
See https://jetbrains.org/intellij/sdk/docs/basics/disposers.html for more details.
The corresponding Disposer.register() stacktrace is shown as the cause:

	at com.intellij.testFramework.TestLoggerFactory$TestLogger.error(TestLoggerFactory.java:400)
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:236)
...
Caused by: java.lang.RuntimeException: Memory leak detected: 'newDisposable' of class com.intellij.openapi.util.Disposer$1 is registered in Disposer but wasn't disposed.
Register it with a proper parentDisposable or ensure that it's always disposed by direct Disposer.dispose call.
See https://jetbrains.org/intellij/sdk/docs/basics/disposers.html for more details.
The corresponding Disposer.register() stacktrace is shown as the cause:

	at com.intellij.openapi.util.ObjectNode.assertNoChildren(ObjectNode.java:44)
```

And:

```
java.lang.NullPointerException: null cannot be cast to non-null type com.intellij.ide.impl.HeadlessDataManager
	at com.intellij.testFramework.TestApplicationManager$Companion.getDataManager(TestApplicationManager.kt:55)
	at com.intellij.testFramework.TestApplicationManager$Companion.access$getDataManager(TestApplicationManager.kt:47)
	at com.intellij.testFramework.TestApplicationManager.setDataProvider(TestApplicationManager.kt:214)
	at com.intellij.testFramework.HeavyPlatformTestCase.initApplication(HeavyPlatformTestCase.java:175)
	at com.intellij.testFramework.HeavyPlatformTestCase.setUp(HeavyPlatformTestCase.java:221)
	at com.intellij.testFramework.HeavyPlatformTestCase.lambda$runBareImpl$23(HeavyPlatformTestCase.java:551)
	at com.intellij.testFramework.EdtTestUtil.lambda$runInEdtAndWait$1(EdtTestUtil.java:40)
```

File: src/test/java/de/cronn/validation_files_diff/impl/GoToValidationFileTest.java (hunk 2 above)
Comment (kevin-keul) [reply src/test/java/de/cronn/validation_files_diff/impl/GoToValidationFileTest.java]: Background why this wasn't working is that it is somewhat a "mixed" strategy in creating validation file names. The main part is done with directory strategy (with "/" delimiter), while the suffix part is done with underscore strategy (with "_" delimiter). The better idea would be to add the different strategies already in the validation assertions library (where at the moment only the underscore strategy is included), but this is out of the scope of this pull request.

It is changed now in such a way that also mixing both strategies is allowed. See ce47e5182dac7d5a73b93df82cbcca5d0ed8883d

**Comment by kevin-keul** [comment]
About https://github.com/cronn/validation-files-comparison-intellij-plugin/pull/21#issuecomment-2172554905 above:

> On my system, the tests do not run successfully. (It works for the master branch though)
> 
> ```
> $ java -version
> openjdk version "17.0.4" 2022-07-19
> ```
> 
> ```
> ./gradlew test
> ```
> 
> failing tests: 
> 
> ...


This is actually two separate problems. 

1) Concerning `NullPointerException`: The `ValidationFileAutomaticRenamerTest` uses `HeavyPlatformTestCase`, which is not working together with the new Abstract Test setup introduced in a future PR and already in parts cherry-picked in bb1950a7a85ea8209618ec5f7671172d3a02c3c0. This is fixed by using the new `AbstractValidationFilePluginTest` instead of `HeavyPlatformTestCase` in 24618e4445877d6e23a604d86334b1169e0b2561. 

2) Concerning `Memory leak detected`: By introducing the new `AbstractValidationFilePluginTest`, an additional check for memory leaks in the tests is performed, which wasn't there before (due to using annotation `@TestApplication`). While the new tests don't have memory leaks, the already existing ones actually do. When running gradle task for tests, all tests are executed and the Memory leaks in existing tests (which appear because of a missing disposing step) become visible. They are fixed in f14aced6d7fad5b8d9704516bb29f5faa88e42a5.

File: README.md (hunk 1 above)
Comment (kevin-keul) [reply README.md:49]: Adjusted as documented in `README.md` in 9c165260b6b3943910729690d9cc5d7aab4427df

File: src/test/java/de/cronn/validation_files_diff/impl/GoToValidationFileTest.java (hunk 2 above)
Comment (pfhartma) [reply src/test/java/de/cronn/validation_files_diff/impl/GoToValidationFileTest.java]: Great, it works well! :slightly_smiling_face:

**Review by pfhartma — APPROVED** [review]
I have no experience with Intellij pugin development, but looks good, as far as I can tell. :slightly_smiling_face:

**Pull request merged**